actix-rt/src/builder.rs

```diff
@@ -127,6 +151,48 @@ impl Builder {
     }
 }
 
+#[derive(Debug)]
+pub struct AsyncSystemRunner {
+    stop: Receiver<i32>,
+    system: System,
+}
+
+impl AsyncSystemRunner {
+    /// This function will start event loop and returns a future that
+    /// resolves once the `System::stop()` function is called.
+    pub fn run_nonblocking(self) -> Box<Future<Item = (), Error = io::Error> + Send + 'static> {
+        let AsyncSystemRunner { stop, .. } = self;
+
+        // run loop
+        Box::new(future::ok(())
```

Comment:
i think you can collapse most of .and_then and put everything into `future::lazy()`